packages/rest/test/unit/coercion/paramStringToNumber.unit.ts

```diff
@@ -3,98 +3,107 @@
   context('valid values', () => {
-    test(['0', NUMBER_SCHEMA_REQUIRED, '0', 0]);
-    test(['1', NUMBER_SCHEMA_REQUIRED, '1', 1]);
-    test(['-1', NUMBER_SCHEMA_REQUIRED, '-1', -1]);
+    test<number>(REQUIRED_NUMBER_PARAM, '0', 0);
+    test<number>(REQUIRED_NUMBER_PARAM, '1', 1);
+    test<number>(REQUIRED_NUMBER_PARAM, '-1', -1);
   });
 
   context('empty values trigger ERROR_BAD_REQUEST', () => {
     // null, '' sent from request are converted to raw value ''
-    test(['empty string', NUMBER_SCHEMA_REQUIRED, '', ERROR_BAD_REQUEST, true]);
+    const errorMsg = HttpErrorMessage.MISSING_REQUIRED(
+      REQUIRED_NUMBER_PARAM.name,
+    );
+    test<HttpErrors.HttpError>(
+      REQUIRED_NUMBER_PARAM,
+      '',
+      new HttpErrors['400'](errorMsg),
+    );
   });
 });
 
 describe('coerce param from string to number - optional', () => {
-  context('valid values', () => {
-    test(['0', NUMBER_SCHEMA, '0', 0]);
-    test(['1', NUMBER_SCHEMA, '1', 1]);
-    test(['-1', NUMBER_SCHEMA, '-1', -1]);
-    test(['1.2', NUMBER_SCHEMA, '1.2', 1.2]);
-    test(['-1.2', NUMBER_SCHEMA, '-1.2', -1.2]);
+  context('valid values', async () => {
+    test<number>(NUMBER_PARAM, '0', 0);
+    test<number>(NUMBER_PARAM, '1', 1);
+    test<number>(NUMBER_PARAM, '-1', -1);
+    test<number>(NUMBER_PARAM, '1.2', 1.2);
+    test<number>(NUMBER_PARAM, '-1.2', -1.2);
   });
 
   context('numbers larger than MAX_SAFE_INTEGER get trimmed', () => {
-    test([
-      'positive large number',
-      NUMBER_SCHEMA,
-      '2343546576878989879789',
-      2.34354657687899e21,
-    ]);
-    test([
-      'negative large number',
-      NUMBER_SCHEMA,
-      '-2343546576878989879789',
-      -2.34354657687899e21,
-    ]);
+    test<number>(NUMBER_PARAM, '2343546576878989879789', 2.34354657687899e21);
+    test<number>(NUMBER_PARAM, '-2343546576878989879789', -2.34354657687899e21);
   });
 
   context('scientific notations', () => {
-    test(['positive number', NUMBER_SCHEMA, '1.234e+30', 1.234e30]);
-    test(['negative number', NUMBER_SCHEMA, '-1.234e+30', -1.234e30]);
+    test<number>(NUMBER_PARAM, '1.234e+30', 1.234e30);
+    test<number>(NUMBER_PARAM, '-1.234e+30', -1.234e30);
   });
 
   context('empty value converts to undefined', () => {
     // [], {} sent from request are converted to raw value undefined
-    test(['empty value as undefined', NUMBER_SCHEMA, undefined, undefined]);
+    test<undefined>(NUMBER_PARAM, undefined, undefined);
   });
 
   context('All other non-number values trigger ERROR_BAD_REQUEST', () => {
+    let errorMsg = HttpErrorMessage.INVALID_DATA('text', NUMBER_PARAM.name);
     // 'false', false, 'true', true, 'text' sent from request are convert to a string
-    test([
-      'invalid value as string',
-      NUMBER_SCHEMA,
+    test<HttpErrors.HttpError>(
+      NUMBER_PARAM,
       'text',
-      ERROR_BAD_REQUEST,
-      true,
-    ]);
+      new HttpErrors['400'](errorMsg),
```

Comment:
+1 for extracting the error message construction to a shared helper.

I am proposing to go one step further and move the creation of HttpError instances to the helper too. 

Before:

```ts
const errorMsg = HttpErrorMessage.INVALID_DATA('text', NUMBER_PARAM.name);
new HttpErrors.BadRequest(errorMsg);
```

After:

```
RestCoercionErrors.invalidData('text', NUMBER_PARAM.name);
```

Implementation:

```ts
export namespace RestCoercionErrors
  export function invalidData<T>(data: T, name: string) => {
    const msg = `Invalid value ${JSON.stringify(data)} for parameter ${name}!`;
    return new HttpErrors.BadRequest(msg);
  }
  // etc.
}
```
    
Thoughts?

Reply:
@bajtos Originally I was thinking of creating templates to unify the message for each kind of http error, not only for coercion functions, that's why I name it as `http-error-message.ts`. I should have moved the file one level up to be under `src`.

It makes more sense that one error message maps to one `HttpError` type, so your approach is better 👍